hub/api/tensor.py

```diff
@@ -25,29 +33,49 @@ def __init__(
             key (str): The internal identifier for this tensor.
             provider (StorageProvider): The storage provider for the parent dataset.
             tensor_slice (slice): The slice object restricting the view of this tensor.
+            tensor_meta (dict): For internal use only. If a tensor with `key` doesn't exist, a new tensor is created with this meta.
+
+        Raises:
+            TensorDoesNotExistError: If no tensor with `key` exists and a `tensor_meta` was not provided.
         """
         self.key = key
         self.provider = provider
         self.slice = tensor_slice
 
-        self.load_meta()
+        if not tensor_exists(self.key, self.provider):
+            if tensor_meta is None:
+                raise TensorDoesNotExistError(self.key)
+
+            create_tensor(self.key, self.provider, tensor_meta)
 
-    def load_meta(self):
-        meta = read_tensor_meta(self.key, self.provider)
-        self.num_samples = meta["length"]
-        self.shape = meta["max_shape"]
+    def append(self, array: np.ndarray, batched: bool):
+        add_samples_to_tensor(
+            array,
+            self.key,
+            storage=self.provider,
+            batched=batched,
+        )
+
+    @property
+    def meta(self):
+        return read_tensor_meta(self.key, self.provider)
+
+    @property
+    def shape(self):
+        # TODO: when dynamic arrays are supported, handle `min_shape != max_shape` (right now they're always equal)
```

Comment:
We can probably just have:
```
tuple(i if i == j else None for i,j in zip(min_shape, max_shape))
```

Reply:
yeah but if we get a `None` for the dimension this could be problematic. i would like it if this shape property could tell me more information.

im almost thinking in the API we could have a `DynamicShape` class (like Index), where when you do `print(shape)` it will print `(10, 5:103, 2:5, 5)` or something. but when you try to use the shape, maybe it acts like a normal tuple where `(10, None, None, 5)`